src/urllib3/connection.py

```diff
@@ -478,125 +483,161 @@ def connect(self) -> None:
+        ssl_sock, self.proxy_is_verified = _ssl_wrap_socket_and_match_hostname(
+            sock,
+            cert_reqs=self.cert_reqs,
+            ssl_version=self.ssl_version,
+            ssl_minimum_version=self.ssl_minimum_version,
+            ssl_maximum_version=self.ssl_maximum_version,
             ca_certs=self.ca_certs,
             ca_cert_dir=self.ca_cert_dir,
             ca_cert_data=self.ca_cert_data,
             server_hostname=hostname,
             ssl_context=ssl_context,
+            # Features that aren't implemented for proxies yet:
+            assert_fingerprint=None,
+            assert_hostname=None,
+            cert_file=None,
+            key_file=None,
+            key_password=None,
+            tls_in_tls=False,
+        )
+        return ssl_sock  # type: ignore[return-value]
+
+
+def _ssl_wrap_socket_and_match_hostname(
+    sock: socket.socket,
+    *,
+    cert_reqs: Union[None, str, int],
+    ssl_version: Union[None, str, int],
+    ssl_minimum_version: Optional[int],
+    ssl_maximum_version: Optional[int],
+    cert_file: Optional[str],
+    key_file: Optional[str],
+    key_password: Optional[str],
+    ca_certs: Optional[str],
+    ca_cert_dir: Optional[str],
+    ca_cert_data: Union[None, str, bytes],
+    assert_hostname: Union[None, str, "Literal[False]"],
+    assert_fingerprint: Optional[str],
+    server_hostname: Optional[str],
+    ssl_context: Optional["ssl.SSLContext"],
+    tls_in_tls: bool = False,
+) -> Tuple[Union["ssl.SSLSocket", "SSLTransport"], bool]:
+    """Logic for constructing an SSLContext from all TLS parameters, passing
+    that down into ssl_wrap_socket, and then doing certificate verification
+    either via hostname or fingerprint. This function exists to guarantee
+    that both proxies and targets have the same behavior when connecting via TLS.
+    """
+
+    # Wrap socket using verification with the root certs in
+    # trusted_root_certs
+    default_ssl_context = False
+    if ssl_context is None:
+        default_ssl_context = True
+        context = create_urllib3_context(
+            ssl_version=resolve_ssl_version(ssl_version),
+            ssl_minimum_version=ssl_minimum_version,
+            ssl_maximum_version=ssl_maximum_version,
+            cert_reqs=resolve_cert_reqs(cert_reqs),
         )
+    else:
+        context = ssl_context
+
+    context.verify_mode = resolve_cert_reqs(cert_reqs)
+
+    # In some cases, we want to verify hostnames ourselves
+    if (
+        # `ssl` can't verify fingerprints or alternate hostnames
+        assert_fingerprint
+        or assert_hostname
+        # We still support OpenSSL 1.0.2, which prevents us from verifying
+        # hostnames easily: https://github.com/pyca/pyopenssl/pull/933
+        or ssl_.IS_PYOPENSSL
+        or not ssl_.HAS_NEVER_CHECK_COMMON_NAME
+    ):
+        context.check_hostname = False
+
+    # Try to load OS default certs if none are given.
+    # Works well on Windows.
```

Comment:
Any reason we explicitly mention this? I would have assumed it worked in Windows without the comment.

Reply:
I think back when it was originally added (by me iirc) the function itself mentioned that it mostly "worked good on Windows". Not sure why I felt like I had to copy that over though. I think it's good to remove and is self-explanatory.